---
title: "DE with complex designs using devil"
output: html_document
vignette: >
  %\VignetteIndexEntry{DE with complex designs using devil}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    fig.width = 7,
    fig.height = 5
)
```

```{r message=FALSE}
library(devil)
library(scRNAseq)
library(SingleCellExperiment)
library(Matrix)
library(dplyr)
library(ggplot2)
```


In this vignette we illustrate how to use **devil** with a **non-trivial design matrix** including multiple covariates and interactions, using the `RichardTCellData()` dataset from the **scRNAseq** package.

## Loading the dataset


```{r}
sce <- scRNAseq::RichardTCellData()
sce
colnames(colData(sce))
head(as.data.frame(colData(sce))[, c("individual", "age", "stimulus", "time", "single.cell.quality")])
```

## Basic filtering

We keep only cells flagged as high quality and do light gene filtering for speed.

```{r}
# Keep only good-quality cells
sce <- sce[, sce$`single cell quality` == "OK"]

# Make sure key covariates are factors
sce$individual <- factor(sce$individual)
sce$age <- factor(sce$age) # e.g. "young"/"old"
sce$stimulus <- droplevels(factor(sce$stimulus))
sce$time <- as.numeric(sce$time)
sce$time_factor <- factor(sce$time, levels = sort(unique(sce$time)))

# Quick summary
table(sce$stimulus, sce$time_factor)
table(sce$age)
length(unique(sce$individual))
```


Filter genes with very low counts (optional, mainly for speed):

```{r}
counts <- assay(sce, "counts")

keep_genes <- Matrix::rowSums(counts > 0) >= 10
sce <- sce[keep_genes, ]
sce <- sce[, sce$stimulus != "unstimulated"]
dim(sce)
```


For this vignettes, we subsample further so that `devil` runs very quickly:

```{r}
set.seed(123)

n_genes <- min(2000, nrow(sce))
n_cells <- min(3000, ncol(sce))

gene_idx <- sample(seq_len(nrow(sce)), n_genes)
cell_idx <- sample(seq_len(ncol(sce)), n_cells)

sce_sub <- sce[gene_idx, cell_idx]
sce_sub
```

From now on weâ€™ll work with `sce_sub`.

## Building a complex design matrix

We consider a model with:

- main effects: `stimulus`, `time_factor`
- interaction: `stimulus:time_factor`
- covariates: `age`


```{r}
meta_df <- as.data.frame(colData(sce_sub))

meta_df$stimulus <- droplevels(factor(meta_df$stimulus))
meta_df$time_factor <- factor(meta_df$time_factor)
meta_df$age <- factor(meta_df$age)
meta_df$individual <- factor(meta_df$individual)

design <- model.matrix(
    ~ stimulus * time_factor,
    data = meta_df
)

head(design)
colnames(design)
```

## Fitting the devil model

Now, `fit_devil()` takes a count matrix and a design matrix and returns coefficients and overdispersions.


```{r}
Y <- as.matrix(assay(sce_sub, "counts"))

devil_fit <- devil::fit_devil(
    input_matrix = Y,
    design_matrix = design,
    verbose = TRUE,
    init_beta_rough = FALSE,
    offset = 1,
    size_factors = "normed_sum"
)

devil_fit
```
