<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>devil statistical model, CPU and GPU implementation ‚Ä¢ devil</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="‚Äùimage/svg+xml‚Äù" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="devil statistical model, CPU and GPU implementation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">devil</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/DE_with_multiple_features.html">Using devil for complex designs with multiple covariates</a></li>
    <li><a class="dropdown-item" href="../articles/devil_GPU_v_CPU.html">devil statistical model, CPU and GPU implementation</a></li>
    <li><a class="dropdown-item" href="../articles/getStarted.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/gpuSetup.html">GPU and CUDA Setup</a></li>
    <li><a class="dropdown-item" href="../articles/quickSetup.html">Quick Setup</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/caravagnalab/devil/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>devil statistical model, CPU and GPU implementation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/devil/vignettes/devil_GPU_v_CPU.Rmd" class="external-link"><code>vignettes/devil_GPU_v_CPU.Rmd</code></a></small>
      <div class="d-none name"><code>devil_GPU_v_CPU.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://caravagnalab.github.io/devil/">devil</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <strong>devil</strong> package implements negative binomial
regression for count data, particularly designed for single-cell RNA-seq
analysis. This vignette explains:</p>
<ol style="list-style-type: decimal">
<li>The statistical model and fitting procedure</li>
<li>Shared computational steps (CPU and GPU)</li>
<li>CPU implementation with practical examples</li>
<li>GPU implementation (conceptual overview)</li>
<li>When to use CPU vs GPU</li>
<li>Performance optimization strategies</li>
</ol>
<div class="section level3">
<h3 id="the-statistical-model">The Statistical Model<a class="anchor" aria-label="anchor" href="#the-statistical-model"></a>
</h3>
<p>Devil fits a <strong>negative binomial generalized linear
model</strong> for each gene:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>‚àº</mo><mtext mathvariant="normal">NB</mtext><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>,</mo><msub><mi>Œ∏</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">
Y_{ij} \sim \text{NB}(\mu_{ij}, \theta_i)
</annotation></semantics></math></p>
<p>where:</p>
<ul>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">Y_{ij}</annotation></semantics></math>
is the count for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
in cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>=</mo><msub><mi>s</mi><mi>j</mi></msub><mo>exp</mo><mrow><mo stretchy="true" form="prefix">(</mo><msubsup><mi>ùêó</mi><mi>j</mi><mi>T</mi></msubsup><msub><mi>ùõÉ</mi><mi>i</mi></msub><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">\mu_{ij} = s_j \exp(\mathbf{X}_j^T \boldsymbol{\beta}_i)</annotation></semantics></math>
is the expected count</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>s</mi><mi>j</mi></msub><annotation encoding="application/x-tex">s_j</annotation></semantics></math>
is the size factor for cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
(normalization)</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ùêó</mi><mi>j</mi></msub><annotation encoding="application/x-tex">\mathbf{X}_j</annotation></semantics></math>
is the design matrix row for cell
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
(covariates)</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ùõÉ</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\boldsymbol{\beta}_i</annotation></semantics></math>
is the coefficient vector for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œ∏</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\theta_i</annotation></semantics></math>
is the overdispersion parameter for gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
</li>
</ul>
<p>The model accounts for:</p>
<ul>
<li>
<strong>Technical variation</strong> via size factors</li>
<li>
<strong>Biological covariates</strong> via the design matrix</li>
<li>
<strong>Overdispersion</strong> beyond Poisson variation</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="common-steps-cpu-and-gpu">Common Steps (CPU and GPU)<a class="anchor" aria-label="anchor" href="#common-steps-cpu-and-gpu"></a>
</h2>
<p>Both implementations share the following preprocessing steps:</p>
<div class="section level3">
<h3 id="size-factor-calculation">1. Size Factor Calculation<a class="anchor" aria-label="anchor" href="#size-factor-calculation"></a>
</h3>
<p>Size factors normalize for differences in sequencing depth across
cells.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate example data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">n_genes</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fl">200</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span><span class="va">n_genes</span> <span class="op">*</span> <span class="va">n_cells</span>, mu <span class="op">=</span> <span class="fl">20</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>    nrow <span class="op">=</span> <span class="va">n_genes</span>, ncol <span class="op">=</span> <span class="va">n_cells</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_genes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cell"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_cells</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create design matrix (two groups)</span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treatment"</span><span class="op">)</span>, each <span class="op">=</span> <span class="va">n_cells</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">group</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="available-normalization-methods">Available normalization methods:<a class="anchor" aria-label="anchor" href="#available-normalization-methods"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Method 1: Geometric mean normalization (default, fast)</span></span>
<span><span class="va">fit_normed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>    input_matrix <span class="op">=</span> <span class="va">counts</span>,</span>
<span>    design_matrix <span class="op">=</span> <span class="va">design</span>,</span>
<span>    size_factors <span class="op">=</span> <span class="st">"normed_sum"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 2: Psi-normalization (robust to highly variable genes)</span></span>
<span><span class="va">fit_psi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>    input_matrix <span class="op">=</span> <span class="va">counts</span>,</span>
<span>    design_matrix <span class="op">=</span> <span class="va">design</span>,</span>
<span>    size_factors <span class="op">=</span> <span class="st">"psinorm"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 3: No normalization (all size factors = 1)</span></span>
<span><span class="va">fit_no_norm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>    input_matrix <span class="op">=</span> <span class="va">counts</span>,</span>
<span>    design_matrix <span class="op">=</span> <span class="va">design</span>,</span>
<span>    size_factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compare size factors</span></span>
<span><span class="va">sf_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_cells</span><span class="op">)</span>,</span>
<span>    normed_sum <span class="op">=</span> <span class="va">fit_normed</span><span class="op">$</span><span class="va">size_factors</span>,</span>
<span>    psinorm <span class="op">=</span> <span class="va">fit_psi</span><span class="op">$</span><span class="va">size_factors</span>,</span>
<span>    no_norm <span class="op">=</span> <span class="va">fit_no_norm</span><span class="op">$</span><span class="va">size_factors</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">sf_comparison</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">cell</span>, names_to <span class="op">=</span> <span class="st">"method"</span>, values_to <span class="op">=</span> <span class="st">"size_factor"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cell</span>, y <span class="op">=</span> <span class="va">size_factor</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html" class="external-link">geom_line</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Size factors across normalization methods"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Cell index"</span>, y <span class="op">=</span> <span class="st">"Size factor"</span><span class="op">)</span></span></code></pre></div>
<p><img src="devil_GPU_v_CPU_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" width="672"></p>
<p><strong>Interpretation</strong>:</p>
<ul>
<li>
<code>normed_sum</code>: Fast, works for most datasets</li>
<li>
<code>psinorm</code>: More robust when many genes are highly
variable</li>
<li>
<code>NULL</code>: Use when data is already normalized</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="offset-vector-computation">2. Offset Vector Computation<a class="anchor" aria-label="anchor" href="#offset-vector-computation"></a>
</h3>
<p>The offset vector incorporates size factors into the model:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The offset is computed as: log(size_factor) + offset_constant</span></span>
<span><span class="va">offset_constant</span> <span class="op">&lt;-</span> <span class="fl">0</span>  <span class="co"># default</span></span>
<span><span class="va">offset_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">fit_normed</span><span class="op">$</span><span class="va">size_factors</span><span class="op">)</span> <span class="op">+</span> <span class="va">offset_constant</span></span>
<span></span>
<span><span class="co"># Visualize the relationship</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    size_factor <span class="op">=</span> <span class="va">fit_normed</span><span class="op">$</span><span class="va">size_factors</span>,</span>
<span>    offset <span class="op">=</span> <span class="va">fit_normed</span><span class="op">$</span><span class="va">offset_vector</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">size_factor</span>, y <span class="op">=</span> <span class="va">offset</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, se <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Relationship between size factors and offsets"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Size factor"</span>, y <span class="op">=</span> <span class="st">"Offset (log scale)"</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="devil_GPU_v_CPU_files/figure-html/unnamed-chunk-5-1.png" class="r-plt" width="672"></p>
</div>
</div>
<div class="section level2">
<h2 id="cpu-implementation">CPU Implementation<a class="anchor" aria-label="anchor" href="#cpu-implementation"></a>
</h2>
<p>The CPU implementation processes genes in parallel using multiple
cores.</p>
<div class="section level3">
<h3 id="workflow-overview">Workflow Overview<a class="anchor" aria-label="anchor" href="#workflow-overview"></a>
</h3>
<pre><code>1. Initialize overdispersion parameters (Œ∏)
   ‚îú‚îÄ If init_overdispersion = NULL: estimate from data
   ‚îî‚îÄ Otherwise: use provided value
   
2. Initialize beta coefficients (Œ≤)
   ‚îú‚îÄ If init_beta_rough = TRUE: simple log-mean initialization
   ‚îî‚îÄ Otherwise: proper GLM initialization
   
3. Fit beta coefficients (parallel across genes)
   
4. Fit overdispersion parameters (parallel across genes)
   ‚îú‚îÄ "MOM": Method of moments (fast, non-iterative)
   ‚îî‚îÄ "old"/"MLE": Maximum likelihood with Cox-Reid adjustment</code></pre>
</div>
<div class="section level3">
<h3 id="cpu-example-with-detailed-timing">CPU Example with Detailed Timing<a class="anchor" aria-label="anchor" href="#cpu-example-with-detailed-timing"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Small dataset for demonstration</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">456</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n_genes_small</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">n_cells_small</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span></span>
<span><span class="co"># Gene-specific baseline expression (log-normal is realistic for RNA-seq)</span></span>
<span><span class="va">gene_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Lognormal.html" class="external-link">rlnorm</a></span><span class="op">(</span></span>
<span>  <span class="va">n_genes_small</span>,</span>
<span>  meanlog <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span>,</span>
<span>  sdlog   <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Gene-specific overdispersion (theta)</span></span>
<span><span class="va">gene_theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Lognormal.html" class="external-link">rlnorm</a></span><span class="op">(</span></span>
<span>  <span class="va">n_genes_small</span>,</span>
<span>  meanlog <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>  sdlog   <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fl">0</span>,</span>
<span>  nrow <span class="op">=</span> <span class="va">n_genes_small</span>,</span>
<span>  ncol <span class="op">=</span> <span class="va">n_cells_small</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">g</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_genes_small</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">counts_small</span><span class="op">[</span><span class="va">g</span>, <span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span></span>
<span>    <span class="va">n_cells_small</span>,</span>
<span>    mu   <span class="op">=</span> <span class="va">gene_means</span><span class="op">[</span><span class="va">g</span><span class="op">]</span>,</span>
<span>    size <span class="op">=</span> <span class="va">gene_theta</span><span class="op">[</span><span class="va">g</span><span class="op">]</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_genes_small</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Cell"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_cells_small</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Two-group design</span></span>
<span><span class="va">design_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treatment"</span><span class="op">)</span>, each <span class="op">=</span> <span class="va">n_cells_small</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># CPU fit with different overdispersion strategies</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">fit_cpu_mom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>        input_matrix <span class="op">=</span> <span class="va">counts_small</span>,</span>
<span>        design_matrix <span class="op">=</span> <span class="va">design_small</span>,</span>
<span>        size_factors <span class="op">=</span> <span class="st">"normed_sum"</span>,</span>
<span>        overdispersion <span class="op">=</span> <span class="st">"MOM"</span>,</span>
<span>        parallel.cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; Compute size factors</span></span>
<span><span class="co">#&gt; Calculating size factors using method: normed_sum</span></span>
<span><span class="co">#&gt; Size factors calculated successfully.</span></span>
<span><span class="co">#&gt; Range: [0.6759, 1.6736]</span></span>
<span><span class="co">#&gt; Initialize theta</span></span>
<span><span class="co">#&gt; Initialize beta</span></span>
<span><span class="co">#&gt; Fitting beta coefficients</span></span>
<span><span class="co">#&gt; Fit overdispersion (mode = MOM)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.075   0.007   0.080</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">fit_cpu_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>        input_matrix <span class="op">=</span> <span class="va">counts_small</span>,</span>
<span>        design_matrix <span class="op">=</span> <span class="va">design_small</span>,</span>
<span>        size_factors <span class="op">=</span> <span class="st">"normed_sum"</span>,</span>
<span>        overdispersion <span class="op">=</span> <span class="st">"MLE"</span>,</span>
<span>        parallel.cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; Compute size factors</span></span>
<span><span class="co">#&gt; Calculating size factors using method: normed_sum</span></span>
<span><span class="co">#&gt; Size factors calculated successfully.</span></span>
<span><span class="co">#&gt; Range: [0.6759, 1.6736]</span></span>
<span><span class="co">#&gt; Initialize theta</span></span>
<span><span class="co">#&gt; Initialize beta</span></span>
<span><span class="co">#&gt; Fitting beta coefficients</span></span>
<span><span class="co">#&gt; Fit overdispersion (mode = MLE)</span></span>
<span><span class="co">#&gt;    user  system elapsed </span></span>
<span><span class="co">#&gt;   0.303   0.432   0.237</span></span></code></pre></div>
<div class="section level4">
<h4 id="examining-iteration-counts">Examining Iteration Counts<a class="anchor" aria-label="anchor" href="#examining-iteration-counts"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Beta fitting iterations (per gene)</span></span>
<span><span class="va">beta_iter_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    iterations <span class="op">=</span> <span class="va">fit_cpu_new</span><span class="op">$</span><span class="va">iterations</span><span class="op">$</span><span class="va">beta_iters</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">beta_iter_summary</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">iterations</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">20</span>, fill <span class="op">=</span> <span class="st">"steelblue"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Beta coefficient fitting: iterations per gene"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"Number of iterations"</span>, y <span class="op">=</span> <span class="st">"Number of genes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="devil_GPU_v_CPU_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" width="672"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Theta fitting iterations (per gene)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">fit_cpu_new</span><span class="op">$</span><span class="va">iterations</span><span class="op">$</span><span class="va">theta_iters</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">theta_iter_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        iterations <span class="op">=</span> <span class="va">fit_cpu_new</span><span class="op">$</span><span class="va">iterations</span><span class="op">$</span><span class="va">theta_iters</span></span>
<span>    <span class="op">)</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">theta_iter_summary</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">iterations</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">20</span>, fill <span class="op">=</span> <span class="st">"coral"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Overdispersion fitting: iterations per gene"</span>,</span>
<span>             x <span class="op">=</span> <span class="st">"Number of iterations"</span>, y <span class="op">=</span> <span class="st">"Number of genes"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="overdispersion-estimation-strategies">Overdispersion Estimation Strategies<a class="anchor" aria-label="anchor" href="#overdispersion-estimation-strategies"></a>
</h3>
<p>The CPU implementation offers three strategies for estimating
overdispersion:</p>
<div class="section level4">
<h4 id="method-of-moments-mom---fastest">1. Method of Moments (MOM) - Fastest<a class="anchor" aria-label="anchor" href="#method-of-moments-mom---fastest"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_mom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>    input_matrix <span class="op">=</span> <span class="va">counts_small</span>,</span>
<span>    design_matrix <span class="op">=</span> <span class="va">design_small</span>,</span>
<span>    overdispersion <span class="op">=</span> <span class="st">"MOM"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># MOM is non-iterative (0 iterations)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Theta iterations (MOM):"</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">fit_mom</span><span class="op">$</span><span class="va">iterations</span><span class="op">$</span><span class="va">theta_iters</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Theta iterations (MOM): 0"</span></span></code></pre></div>
<p><strong>When to use</strong>:<br>
Large datasets or GPU-based workflows, when speed and scalability are
critical and a fully iterative dispersion fit is impractical.</p>
<p><strong>How it works</strong>:<br>
In a regression setting, MOM does <strong>not</strong> rely on the raw
sample mean and variance of counts. Instead, it estimates overdispersion
from the <strong>residual variability around the fitted mean</strong>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{ij}</annotation></semantics></math>
implied by the design matrix and offsets.</p>
<p>Concretely, after fitting (or updating) the regression coefficients
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ùõÉ</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\boldsymbol{\beta}_i</annotation></semantics></math>,
devil computes fitted means
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{ij}</annotation></semantics></math>
and matches the negative binomial mean‚Äìvariance relationship
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mi mathvariant="normal">V</mi><mi mathvariant="normal">a</mi><mi mathvariant="normal">r</mi></mrow><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>+</mo><mfrac><msubsup><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow><mn>2</mn></msubsup><msub><mi>Œ∏</mi><mi>i</mi></msub></mfrac></mrow><annotation encoding="application/x-tex">
\mathrm{Var}(Y_{ij}) = \mu_{ij} + \frac{\mu_{ij}^2}{\theta_i}
</annotation></semantics></math> using squared residuals
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><msub><mi>y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo>‚àí</mo><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(y_{ij}-\mu_{ij})^2</annotation></semantics></math>
to obtain a <strong>closed-form, non-iterative</strong> estimate of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œ∏</mi><mi>i</mi></msub><annotation encoding="application/x-tex">\theta_i</annotation></semantics></math>.</p>
<p>This makes MOM extremely fast and well suited for large-scale
single-cell analyses, while still accounting for the effects of
covariates and normalization through
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Œº</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">\mu_{ij}</annotation></semantics></math>.</p>
</div>
<div class="section level4">
<h4 id="mle-with-cox-reid-mle---most-accurate">2. MLE with Cox-Reid (MLE) - Most Accurate<a class="anchor" aria-label="anchor" href="#mle-with-cox-reid-mle---most-accurate"></a>
</h4>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_mle</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>    input_matrix <span class="op">=</span> <span class="va">counts_small</span>,</span>
<span>    design_matrix <span class="op">=</span> <span class="va">design_small</span>,</span>
<span>    overdispersion <span class="op">=</span> <span class="st">"old"</span>,</span>
<span>    max_iter <span class="op">=</span> <span class="fl">200</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>When to use</strong>: Maximum precision, small datasets.
<strong>How it works</strong>: Maximum likelihood estimation with
Cox-Reid adjustment.</p>
</div>
<div class="section level4">
<h4 id="comparing-overdispersion-estimates">Comparing Overdispersion Estimates<a class="anchor" aria-label="anchor" href="#comparing-overdispersion-estimates"></a>
</h4>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">theta_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_genes_small</span><span class="op">)</span>,</span>
<span>    MOM <span class="op">=</span> <span class="va">fit_mom</span><span class="op">$</span><span class="va">overdispersion</span>,</span>
<span>    MLE <span class="op">=</span> <span class="va">fit_mle</span><span class="op">$</span><span class="va">overdispersion</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">theta_comparison</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co">#pivot_longer(-gene, names_to = "method", values_to = "theta") %&gt;%</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">MOM</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">MLE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Overdispersion estimates across methods"</span>,</span>
<span>         x <span class="op">=</span> <span class="st">"log10(Œ∏ MOM)"</span>, y <span class="op">=</span> <span class="st">"log10(Œ∏ MLE)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="devil_GPU_v_CPU_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" width="672"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Correlation between methods</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">theta_comparison</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;     MOM MLE</span></span>
<span><span class="co">#&gt; MOM   1   1</span></span>
<span><span class="co">#&gt; MLE   1   1</span></span></code></pre></div>
<p><strong>Interpretation</strong>: Methods generally agree, but MLE may
provide more stable estimates for difficult genes. MOM is fastest but
may be less precise for genes with extreme overdispersion.</p>
</div>
</div>
<div class="section level3">
<h3 id="beta-initialization-strategies">Beta Initialization Strategies<a class="anchor" aria-label="anchor" href="#beta-initialization-strategies"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compare initialization strategies</span></span>
<span><span class="va">time_rough</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">fit_rough</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>        input_matrix <span class="op">=</span> <span class="va">counts_small</span>,</span>
<span>        design_matrix <span class="op">=</span> <span class="va">design_small</span>,</span>
<span>        init_beta_rough <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">time_proper</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">fit_proper</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span></span>
<span>        input_matrix <span class="op">=</span> <span class="va">counts_small</span>,</span>
<span>        design_matrix <span class="op">=</span> <span class="va">design_small</span>,</span>
<span>        init_beta_rough <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>        verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Rough init:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">time_rough</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"sec"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Rough init: 0.06 sec"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Proper init:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">time_proper</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"sec"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Proper init: 0.07 sec"</span></span>
<span></span>
<span><span class="co"># Check if results are similar</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">fit_rough</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, <span class="va">fit_proper</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.9999904</span></span></code></pre></div>
<p><strong>Recommendation</strong>: -
<code>init_beta_rough = FALSE</code> (default): Proper GLM
initialization, more robust - <code>init_beta_rough = TRUE</code>:
Faster for very large datasets</p>
<p>The rough initialization simply uses <code>log(mean(y) + 1)</code>
for the intercept and zeros for other coefficients. The proper
initialization fits a simplified GLM to get better starting values.</p>
</div>
</div>
<div class="section level2">
<h2 id="gpu-implementation-conceptual">GPU Implementation (Conceptual)<a class="anchor" aria-label="anchor" href="#gpu-implementation-conceptual"></a>
</h2>
<p>The GPU implementation uses CUDA for massive parallelization across
genes.</p>
<div class="section level3">
<h3 id="gpu-workflow">GPU Workflow<a class="anchor" aria-label="anchor" href="#gpu-workflow"></a>
</h3>
<pre><code>1. Compute size factors and offsets (same as CPU)
   ‚îî‚îÄ This preprocessing happens on CPU

2. Batch genes for memory efficiency
   ‚îú‚îÄ Genes are divided into batches of size 'batch_size'
   ‚îî‚îÄ Default: 1024 genes per batch

3. GPU kernel: Fit beta coefficients
   ‚îú‚îÄ Each GPU thread handles one gene
   ‚îú‚îÄ Thousands of genes fitted simultaneously
   ‚îî‚îÄ Iterative weighted least squares on GPU

4. Overdispersion: Method of Moments only
   ‚îú‚îÄ Computed on GPU (non-iterative)
   ‚îî‚îÄ Fast calculation across all genes</code></pre>
</div>
<div class="section level3">
<h3 id="key-differences-from-cpu">Key Differences from CPU<a class="anchor" aria-label="anchor" href="#key-differences-from-cpu"></a>
</h3>
<table class="table">
<colgroup>
<col width="47%">
<col width="26%">
<col width="26%">
</colgroup>
<thead><tr class="header">
<th>Feature</th>
<th>CPU</th>
<th>GPU</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Parallelization</strong></td>
<td>Across genes using CPU cores (typically 4-64)</td>
<td>Across genes using GPU threads (thousands)</td>
</tr>
<tr class="even">
<td><strong>Memory model</strong></td>
<td>Distributed (per core)</td>
<td>Shared (batched on GPU memory)</td>
</tr>
<tr class="odd">
<td><strong>Overdispersion</strong></td>
<td>MOM, Iterative, or MLE</td>
<td>MOM only</td>
</tr>
<tr class="even">
<td><strong>Beta initialization</strong></td>
<td>Full initialization available</td>
<td>Simplified initialization</td>
</tr>
<tr class="odd">
<td><strong>Batch processing</strong></td>
<td>All genes or chunked by core</td>
<td>Always batched for GPU memory</td>
</tr>
<tr class="even">
<td><strong>Best for</strong></td>
<td>Small-medium datasets (&lt;5k genes)</td>
<td>Large datasets (&gt;10k genes, &gt;5k cells)</td>
</tr>
<tr class="odd">
<td><strong>Setup</strong></td>
<td>Works out of the box</td>
<td>Requires CUDA toolkit and recompilation</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Giovanni Santacatterina.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
