<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction • devil</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">devil</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/caravagnalab/devil/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/devil/vignettes/articles/Introduction.Rmd" class="external-link"><code>vignettes/articles/Introduction.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://caravagnalab.github.io/devil/" class="external-link">devil</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,</span></span>
<span><span class="co">#&gt;     tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span></code></pre></div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>Let’s load a single-cell RNA dataset from the <code>scRNAseq</code>
package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ReprocessedData.html" class="external-link">ReprocessedFluidigmData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>In order to obtain a devil fit you will need the RNA counts and the
metadata describing the single cells’ features.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">@</span><span class="va">colData</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 26255   130</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 130  28</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 28 columns</span></span>
<span><span class="co">#&gt;               NREADS  NALIGNED    RALIGN TOTAL_DUP    PRIMER INSERT_SZ</span></span>
<span><span class="co">#&gt;            &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SRR1275356  10554900   7555880   71.5862   58.4931 0.0217638       208</span></span>
<span><span class="co">#&gt; SRR1274090    196162    182494   93.0323   14.5122 0.0366826       247</span></span>
<span><span class="co">#&gt; SRR1275251   8524470   5858130   68.7213   65.0428 0.0351827       230</span></span>
<span><span class="co">#&gt; SRR1275287   7229920   5891540   81.4884   49.7609 0.0208685       222</span></span>
<span><span class="co">#&gt; SRR1275364   5403640   4482910   82.9609   66.5788 0.0298284       228</span></span>
<span><span class="co">#&gt; SRR1275269  10729700   7806230   72.7536   50.4285 0.0204349       245</span></span>
<span><span class="co">#&gt;            INSERT_SZ_STD COMPLEXITY     NDUPR PCT_RIBOSOMAL_BASES</span></span>
<span><span class="co">#&gt;                &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;           &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SRR1275356            63   0.868928  0.343113               2e-06</span></span>
<span><span class="co">#&gt; SRR1274090           133   0.997655  0.935730               0e+00</span></span>
<span><span class="co">#&gt; SRR1275251            89   0.789252  0.201082               0e+00</span></span>
<span><span class="co">#&gt; SRR1275287            78   0.898100  0.538191               0e+00</span></span>
<span><span class="co">#&gt; SRR1275364            76   0.890693  0.391660               0e+00</span></span>
<span><span class="co">#&gt; SRR1275269            99   0.879414  0.431169               0e+00</span></span>
<span><span class="co">#&gt;            PCT_CODING_BASES PCT_UTR_BASES PCT_INTRONIC_BASES</span></span>
<span><span class="co">#&gt;                   &lt;numeric&gt;     &lt;numeric&gt;          &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SRR1275356         0.125806      0.180954           0.613229</span></span>
<span><span class="co">#&gt; SRR1274090         0.309822      0.412917           0.205185</span></span>
<span><span class="co">#&gt; SRR1275251         0.398461      0.473884           0.039886</span></span>
<span><span class="co">#&gt; SRR1275287         0.196420      0.227592           0.498944</span></span>
<span><span class="co">#&gt; SRR1275364         0.138617      0.210406           0.543941</span></span>
<span><span class="co">#&gt; SRR1275269         0.333077      0.354635           0.248331</span></span>
<span><span class="co">#&gt;            PCT_INTERGENIC_BASES PCT_MRNA_BASES MEDIAN_CV_COVERAGE</span></span>
<span><span class="co">#&gt;                       &lt;numeric&gt;      &lt;numeric&gt;          &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SRR1275356             0.080008       0.306760           1.495770</span></span>
<span><span class="co">#&gt; SRR1274090             0.072076       0.722739           1.007580</span></span>
<span><span class="co">#&gt; SRR1275251             0.087770       0.872345           1.242990</span></span>
<span><span class="co">#&gt; SRR1275287             0.077044       0.424013           0.775981</span></span>
<span><span class="co">#&gt; SRR1275364             0.107035       0.349024           1.441370</span></span>
<span><span class="co">#&gt; SRR1275269             0.063957       0.687712           0.617100</span></span>
<span><span class="co">#&gt;            MEDIAN_5PRIME_BIAS MEDIAN_3PRIME_BIAS MEDIAN_5PRIME_TO_3PRIME_BIAS</span></span>
<span><span class="co">#&gt;                     &lt;numeric&gt;          &lt;numeric&gt;                    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; SRR1275356           0.000000           0.166122                     1.036250</span></span>
<span><span class="co">#&gt; SRR1274090           0.181742           0.698991                     0.293510</span></span>
<span><span class="co">#&gt; SRR1275251           0.000000           0.340046                     0.201518</span></span>
<span><span class="co">#&gt; SRR1275287           0.010251           0.350915                     0.292838</span></span>
<span><span class="co">#&gt; SRR1275364           0.000000           0.204074                     0.619863</span></span>
<span><span class="co">#&gt; SRR1275269           0.057960           0.345502                     0.284480</span></span>
<span><span class="co">#&gt;            sample_id.x           Lane_ID LibraryName avgLength     spots</span></span>
<span><span class="co">#&gt;            &lt;character&gt;       &lt;character&gt; &lt;character&gt; &lt;integer&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; SRR1275356   SRX534610 D24VYACXX130502:4      GW16_2       202   9818076</span></span>
<span><span class="co">#&gt; SRR1274090   SRX534823                 1       NPC_9        60     95454</span></span>
<span><span class="co">#&gt; SRR1275251   SRX534623 D24VYACXX130502:4      GW16_8       202   7935952</span></span>
<span><span class="co">#&gt; SRR1275287   SRX534641 D24VYACXX130502:1    GW21+3_2       202   6531944</span></span>
<span><span class="co">#&gt; SRR1275364   SRX534614 D24VYACXX130502:7     GW16_23       202   4919561</span></span>
<span><span class="co">#&gt; SRR1275269   SRX534632 D24VYACXX130502:4      GW21_8       202   9969377</span></span>
<span><span class="co">#&gt;            Biological_Condition Coverage_Type Cluster1 Cluster2</span></span>
<span><span class="co">#&gt;                     &lt;character&gt;   &lt;character&gt; &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">#&gt; SRR1275356                 GW16          High     IIIb      III</span></span>
<span><span class="co">#&gt; SRR1274090                  NPC           Low     1a        I  </span></span>
<span><span class="co">#&gt; SRR1275251                 GW16          High     NA        III</span></span>
<span><span class="co">#&gt; SRR1275287               GW21+3          High     1c        I  </span></span>
<span><span class="co">#&gt; SRR1275364                 GW16          High     IIIb      III</span></span>
<span><span class="co">#&gt; SRR1275269                 GW21          High     NA        I</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "NREADS"                       "NALIGNED"                    </span></span>
<span><span class="co">#&gt;  [3] "RALIGN"                       "TOTAL_DUP"                   </span></span>
<span><span class="co">#&gt;  [5] "PRIMER"                       "INSERT_SZ"                   </span></span>
<span><span class="co">#&gt;  [7] "INSERT_SZ_STD"                "COMPLEXITY"                  </span></span>
<span><span class="co">#&gt;  [9] "NDUPR"                        "PCT_RIBOSOMAL_BASES"         </span></span>
<span><span class="co">#&gt; [11] "PCT_CODING_BASES"             "PCT_UTR_BASES"               </span></span>
<span><span class="co">#&gt; [13] "PCT_INTRONIC_BASES"           "PCT_INTERGENIC_BASES"        </span></span>
<span><span class="co">#&gt; [15] "PCT_MRNA_BASES"               "MEDIAN_CV_COVERAGE"          </span></span>
<span><span class="co">#&gt; [17] "MEDIAN_5PRIME_BIAS"           "MEDIAN_3PRIME_BIAS"          </span></span>
<span><span class="co">#&gt; [19] "MEDIAN_5PRIME_TO_3PRIME_BIAS" "sample_id.x"                 </span></span>
<span><span class="co">#&gt; [21] "Lane_ID"                      "LibraryName"                 </span></span>
<span><span class="co">#&gt; [23] "avgLength"                    "spots"                       </span></span>
<span><span class="co">#&gt; [25] "Biological_Condition"         "Coverage_Type"               </span></span>
<span><span class="co">#&gt; [27] "Cluster1"                     "Cluster2"</span></span></code></pre></div>
<p>You can see that we have 130 cells described by 28 features along
with the counts of 26255 genes. We remove the non-expressed genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 17085   130</span></span></code></pre></div>
<p>We remain with ~17k genes. Now we need to produce so called
<code>design_matrix</code> which is a matrix that contains a row per
cell and each row contains some of the cell features. To do so, the
simplest way is to use the <code>model.matrix</code> function, which
requires a formula and the metadata as input.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">Biological_Condition</span>, data <span class="op">=</span> <span class="va">metadata</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Biological_Condition</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "GW16"   "NPC"    "GW21+3" "GW21"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">design_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt;            (Intercept) Biological_ConditionGW21 Biological_ConditionGW21+3</span></span>
<span><span class="co">#&gt; SRR1275356           1                        0                          0</span></span>
<span><span class="co">#&gt; SRR1274090           1                        0                          0</span></span>
<span><span class="co">#&gt; SRR1275251           1                        0                          0</span></span>
<span><span class="co">#&gt; SRR1275287           1                        0                          1</span></span>
<span><span class="co">#&gt; SRR1275364           1                        0                          0</span></span>
<span><span class="co">#&gt; SRR1275269           1                        1                          0</span></span>
<span><span class="co">#&gt;            Biological_ConditionNPC</span></span>
<span><span class="co">#&gt; SRR1275356                       0</span></span>
<span><span class="co">#&gt; SRR1274090                       1</span></span>
<span><span class="co">#&gt; SRR1275251                       0</span></span>
<span><span class="co">#&gt; SRR1275287                       0</span></span>
<span><span class="co">#&gt; SRR1275364                       0</span></span>
<span><span class="co">#&gt; SRR1275269                       0</span></span></code></pre></div>
<p>In this case we model each cell considering its
<code>Biological_Condition</code>. In our case, the intercept represent
the “GW16” condition.</p>
</div>
<div class="section level2">
<h2 id="fitting">Fitting<a class="anchor" aria-label="anchor" href="#fitting"></a>
</h2>
<p>We are now ready to use the <code>fit_devil</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">devil</span><span class="fu">::</span><span class="fu"><a href="../reference/fit_devil.html">fit_devil</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>, <span class="va">design_matrix</span>, overdispersion <span class="op">=</span> <span class="cn">T</span>, size_factors <span class="op">=</span> <span class="cn">T</span>, verbose <span class="op">=</span> <span class="cn">T</span>, parallel.cores <span class="op">=</span> <span class="fl">1</span>, min_cells <span class="op">=</span> <span class="fl">100</span>, avg_counts <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; Removing 16492 lowly expressed genes.</span></span>
<span><span class="co">#&gt; Compute size factors</span></span>
<span><span class="co">#&gt; Initialize beta estimate</span></span>
<span><span class="co">#&gt; Fit beta coefficients</span></span>
<span><span class="co">#&gt; Fit overdispersion</span></span></code></pre></div>
<p>The obtained fit contains, along with additional values:</p>
<ul>
<li>
<code>beta</code> : matrix of coefficients (n_genes x
n_features)</li>
<li>
<code>overdispersion</code> : vector of coefficients (n_genes)</li>
</ul>
</div>
<div class="section level2">
<h2 id="testing">Testing<a class="anchor" aria-label="anchor" href="#testing"></a>
</h2>
<p>In order to test the data you need to specify your null hypothesis
using a contrast vector
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>c</mi><annotation encoding="application/x-tex">c</annotation></semantics></math>.
Considering a gene
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>g</mi><annotation encoding="application/x-tex">g</annotation></semantics></math>
along with its inferred coefficient
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>β</mi><mi>g</mi></msub><annotation encoding="application/x-tex">\beta_g</annotation></semantics></math>,
the null hypothesis
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>H</mi><mn>0</mn></msub><annotation encoding="application/x-tex">H_0</annotation></semantics></math>
is usually defined as</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>H</mi><mn>0</mn></msub><mo>:</mo><munder><mo>∑</mo><mi>i</mi></munder><msub><mi>c</mi><mi>i</mi></msub><msub><mi>β</mi><mrow><mi>g</mi><mo>,</mo><mi>i</mi></mrow></msub><mo>=</mo><mn>0</mn></mrow><annotation encoding="application/x-tex"> H_0 : \sum_i c_i \beta_{g,i} = 0</annotation></semantics></math>
For example, if you are interested in the genes that are differentially
expressed between the “GW21” and the “NPC” condition, you need to find
the genes for which we strongly reject the null hypothesis</p>
<p><span class="math display">$$ \beta_{GW21} = \beta_{NPC} \hspace{5mm}
\rightarrow \hspace{5mm} \beta_{GW21} - \beta_{NPC} = 0$$</span> which
is equivalent to defining the contrast vector
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>c</mi><mo>=</mo><mrow><mo stretchy="true" form="prefix">(</mo><mn>0</mn><mo>,</mo><mn>1</mn><mo>,</mo><mn>0</mn><mo>,</mo><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">c = (0,1,0,-1)</annotation></semantics></math>.
Once the contrast vector is defined, you can test the null hypothesis
using the <code>test_de</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">test_res</span> <span class="op">&lt;-</span> <span class="fu">devil</span><span class="fu">::</span><span class="fu"><a href="../reference/test_de.html">test_de</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrast</span>, max_lfc <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="st">'name'</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">test_res</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">test_res</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">test_res</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">test_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "pval"     "adj_pval" "lfc"      "name"</span></span></code></pre></div>
<p>The results contains, for each gene</p>
<ul>
<li>
<code>pval</code> : p-value associated with the statistical
test</li>
<li>
<code>adj_pval</code> : p-value corrected considering multiple
testing</li>
<li>
<code>lfc</code> : log2 fold change of gene expression between the
tested conditions</li>
</ul>
<p>You can also visualize the results using the
<code>EnhancedVolcano</code> package</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devil</span><span class="fu">::</span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">test_res</span>, lfc_cut <span class="op">=</span> <span class="fl">1</span>, pval_cut <span class="op">=</span> <span class="fl">.05</span>, labels <span class="op">=</span> <span class="cn">TRUE</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Introduction_files/figure-html/unnamed-chunk-8-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Giovanni Santacatterina.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
